fs/tty.c

```diff
@@ -414,7 +408,7 @@ static bool pty_is_half_closed_master(struct tty *tty) {
     struct tty *slave = tty->pty.other;
     // only time one tty lock is nested in another
     lock(&slave->lock);
-    bool half_closed = slave->ever_opened && slave->refcount == 1;
+    bool half_closed = slave->ever_opened && (slave->refcount == 1 || slave->hung_up);
```

Comment:
I wonder if the refcount check is still necessary? Would the new pty_slave_close do that?

Reply:
This is needed when `pty_slave_cleanup()` runs, I think.  It calls `pty_hangup_other()`, which calls `tty_hangup()` on the master.  That calls `tty_input_wakeup()` on the master, which may result in that user program waking up in `tty_read()`, or calling `read()` again, before the pty slave device instance is fully cleaned up and the refcount decremented.  That's what I remember, anyway.